Add support for zsh

pinecat opened the pull request:
This patch adds the necesary lines in rust2rpm.toml and atuin.spec to enable atuin support for users running the Z shell (zsh). If the $ZSH_VERSION environment variable has been set (which is done by default upon starting zsh), it will cause zsh to source the file /usr/libexec/atuin/atuin-init.zsh, which contains the code normally generated by 'atuin init zsh'.

The /etc/zprofile script calls 'emulate -L ksh', before sourcing files in /etc/profile.d for compatibility reasons. However, atuin needs some zsh-specific functions (mainly, compinit), so we call 'emulate -L zsh' before running through the code in /usr/libexec/atuin/atuin-init.zsh, and then call 'emulate -L ksh' at the very end.

pinecat committed:
Add support for zsh

This patch adds the necesary lines in rust2rpm.toml and atuin.spec to
enable atuin support for users running the Z shell (zsh). If the
$ZSH_VERSION environment variable has been set (which is done by default
upon starting zsh), it will cause zsh to source the file
/usr/libexec/atuin/atuin-init.zsh, which contains the code normally
generated by 'atuin init zsh'.

The /etc/zprofile script calls 'emulate -L ksh', before sourcing files
in /etc/profile.d for compatibility reasons. However, atuin needs some
zsh-specific functions (mainly, compinit), so we call 'emulate -L zsh'
before running through the code in /usr/libexec/atuin/atuin-init.zsh,
and then call 'emulate -L ksh' at the very end.

pinecat commented:
One thing, the RPM seems to install the `atuin` binary at /bin/atuin/atuin. So, I ran the following command after starting up my container and installing the *.rpm files: `mv /bin/atuin/atuin /tmp && rmdir /bin/atuin && mv /tmp/atuin /bin`. After that, it was all working. I'm not sure if you meant to install atuin here, but if so, we would need to update /usr/libexec/atuin/atuin-init.zsh, since atuin wouldn't be in any $PATH. I have not dived into the `install` commands yet, though. I'm happy to make the relevant changes/updates, so please let me know. Thanks!

LecrisUT commented on atuin/atuin.spec:
```diff
@@ -116,9 +116,30 @@ cat > other_installs/profile.d/atuin.sh <<EOF
+# Check for zsh environment
+if [ -n "\$ZSH_VERSION-}" ]; then
+  # Skip noninteractive shells.
+  [[ \$- != *i* ]] && return
```
Does zsh have this check for non-interactive? This one I took from the bash flags explicitly

pinecat replied:
Yes, if you are interested you can test it yourself with:
```shell
[[ $- == *i* ]] && echo 'Interactive' || echo 'Not interactive' # -> Interactive
```
Or in a file:
```shell
#!/bin/zsh
[[ $- == *i* ]] && echo 'Interactive' || echo 'Not interactive' # -> Non interactive
```

LecrisUT commented on atuin/atuin.spec:
```diff
@@ -116,9 +116,30 @@ cat > other_installs/profile.d/atuin.sh <<EOF
+
+  # Make sure to go back to emulating ksh after we finish
+  # initializing atuin.
+  emulate -L ksh
```
I'm not familiar with the implications of this, but this is set by Fedora default installation or something? Could we move the order of the `atuin.sh` to avoid the `emulate` parts?

pinecat replied:
It is called in `/etc/zprofile`. This is a file that is sourced by zsh automatically when it starts up. Unfortunately, it is sourced before anything in `/etc/profile.d` (to be more specific, it is the file that sources everything in `/etc/profile.d`). Here is my `/etc/zprofile`. It is the standard that ships with all zsh installs on Fedora:
```shell
#
# /etc/zprofile and ~/.zprofile are run for login shells
#

_src_etc_profile()
{
    #  Make /etc/profile happier, and have possible ~/.zshenv options like
    # NOMATCH ignored.
    #
    emulate -L ksh

    # source profile
    if [ -f /etc/profile ]; then
	    source /etc/profile
    fi
}
_src_etc_profile

unset -f _src_etc_profile
```

LecrisUT replied:
Great, I'm convinced :slightly_smiling_face:. Could you put a comment pointing to `/etc/zprofile`?

LecrisUT commented on atuin/atuin.spec:
```diff
@@ -116,9 +116,30 @@ cat > other_installs/profile.d/atuin.sh <<EOF
 fi
+
+# Check for zsh environment
+if [ -n "\$ZSH_VERSION-}" ]; then
```
Can you `elif` this just to keep it a tidy flow

pinecat replied:
Yes, that makes sense. I will get in a v2 patch for you, thanks!

LecrisUT reviewed (commented):
Thanks for the suggestion. `atuin` is under review so I won't push it up there, but I'll add it after the review is finished.

For the `emulate ksh` I could use some more reading material. Want to meake sure we are not changing the uaer setup unexpectedly. Do you have some other `profile.d` or user scripts to reference?

pinecat commented:
> Thanks for the suggestion. `atuin` is under review so I won't push it up there, but I'll add it after the review is finished.
> 
> For the `emulate ksh` I could use some more reading material. Want to meake sure we are not changing the uaer setup unexpectedly. Do you have some other `profile.d` or user scripts to reference?

I'm not actually sure how other `/etc/profile.d` scripts handle this. I will grep around there and see if I can find anything. It seems like I may be able to call `emulate zsh ...` on a single line, rather than switching back and forth, which may be safer. The only official docs I know of on `emulate` are from the manpage: https://linux.die.net/man/1/zshbuiltins.

pinecat committed:
Refactor how zsh's builtin 'emulate' command is used

This patch removes the second 'emulate [-L ksh]' call in favor of
passing the '-c' flag to run a single line of code outside of ksh.
Directly from the zshbuiltins(1) manpage: "If -c arg appears in flags,
arg is evaluated while the requested emulation is  temporarily  in
effect.  In this case the emulation mode and all options are restored to
their previous values before emulate returns."

The check for the zsh environment was also moved to an elif block,
instead of it's own if statement.

pinecat commented:
> > Thanks for the suggestion. `atuin` is under review so I won't push it up there, but I'll add it after the review is finished.
> > For the `emulate ksh` I could use some more reading material. Want to meake sure we are not changing the uaer setup unexpectedly. Do you have some other `profile.d` or user scripts to reference?
> 
> I'm not actually sure how other `/etc/profile.d` scripts handle this. I will grep around there and see if I can find anything. It seems like I may be able to call `emulate zsh ...` on a single line, rather than switching back and forth, which may be safer. The only official docs I know of on `emulate` are from the manpage: https://linux.die.net/man/1/zshbuiltins.

Unfortunately, it doesn't look like I have any other scripts in `/etc/profile.d` that use 'emulate'. It's possible that some exist out there, but I don't have the time today to look into it. I did, however, refactor the command so only a single line is emulated in the zsh mode (the line sourcing `/usr/libexec/atuin/atuin-init.zsh`). I think this solution is a little more elegant. Also, I included it in my commit message, but I'll add it here as well: Directly from the `zshbuiltins(1)` manpage: 
```
If -c arg appears in flags, arg is evaluated 
while the requested emulation is  temporarily 
in effect.  In this case the emulation mode and 
all options are restored to their previous values 
before emulate returns.
```
Finally, the check for the zsh environment was moved to an elif block, instead of it's own if statement. Please let me know if there is anything else you'd like updated. Thank you!

LecrisUT commented:
Looks great, thanks a lot of researching these. I will merge it and add it downstream as soon as [the review](https://bugzilla.redhat.com/show_bug.cgi?id=2297310) is accepted.

pinecat commented:
Not a problem, thanks so much for your help, and for getting this reviewed! I'm looking forward to having atuin in the official repos. If there is anything else I can do, please feel free to reach out. Thanks!
